mono/utils/mono-state.c

```diff
@@ -927,6 +927,63 @@ mono_native_state_add_memory (MonoStateWriter *writer)
 	mono_state_writer_printf(writer, "},\n");
 }
 
+#define MONO_CRASH_REPORTING_MAPPING_LINE_LIMIT 30
+
+static void
+mono_native_state_add_process_map (MonoStateWriter *writer)
+{
+#if defined(__linux__) && !defined(HOST_ANDROID)
+	int handle = g_open ("/proc/self/maps", O_RDONLY, S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH);
+	if (handle == -1) {
+		g_async_safe_printf ("Couldn't find /proc/self/maps on Linux system. Continuing.");
+		return;
+	}
+
+	assert_has_space (writer);
+	mono_state_writer_indent (writer);
+	mono_state_writer_object_key (writer, "process_map");
+	mono_state_writer_printf(writer, "[\n");
+
+	int mapping = 0;
+	while (mapping < MONO_CRASH_REPORTING_MAPPING_LINE_LIMIT) {
```

Comment:
`MONO_CRASH_REPORTING_MAPPING_LINE_LIMIT` shouldn't be there (or maybe something crazy like 9999). and it doesn't seem to handle EOF; what if `/proc/self/maps` has fewer than `MONO_CRASH_REPORTING_MAPPING_LINE_LIMIT` entries?

Reply:
So there's a couple points in there, let me handle them one by one.

``` 
MONO_CRASH_REPORTING_MAPPING_LINE_LIMIT shouldn't be there (or maybe something crazy like 9999)
```

We had a line limit on the old printer, and it seemed even more important here to limit it, where the crash output writer doesn't have realloc right now. 

I think you can definitely bump the constant if you're not seeing enough info for MonoDevelop or other large apps, but it should be a constant (not a magic number in the source) and it should probably be a bit smaller than 9999. 

```
what if /proc/self/maps has fewer than MONO_CRASH_REPORTING_MAPPING_LINE_LIMIT entries?
```

That's the common case on my machine during testing, where it only prints a couple of lines for a "GC.GetGeneration" example with an assertion in the icall. It works fine for me. 

You should remember that it continues until `read` fails, in the underlying reader. So in this case, it just prints a few zero-length string entries at the end of the array. 

If you're concerned about the space usage there, feel free to make it end earlier. This seems like an edge case only hit by very small programs that fail very quickly, and it doesn't fail in a way that leads to invalid JSON or anything breaking.

Reply:
> We had a line limit on the old printer

In the beginning we didn't have the limit, but `/proc/self/maps` can be large, thus we limited for user ease. It effectively rendered it useless because you really want to have the full content of `maps` available in order to debug an issue... not the first few mappings. Thus I suggested to completely drop it.

Having a partial snapshot of `/proc/self/maps` in the JSON file is useless too. We really want a full mapping (or drop it from JSON too).

What causes `realloc` here? `mono_state_writer_printf (writer, "...")`?

Reply:
I meant to say that it's possible to run out of space: https://github.com/mono/mono/blob/master/mono/utils/mono-state.c#L368

So we don't want to be fastidious everywhere else and then allow ourselves a few hundred lines of `/proc/self/maps`. There should probably be some kind of limit. Or maybe a bit of parsing.